lib/primitive/lib/Cardano/Wallet/Primitive/Types/DRep.hs

```diff
@@ -41,74 +49,80 @@ import GHC.Generics
     deriving (Eq, Generic, Show, Ord)
     deriving anyclass NFData
 
--- | Encode 'DRepKeyHash' as Bech32 with "drep" hrp.
-encodeDRepKeyHashBech32 :: DRepKeyHash -> Text
-encodeDRepKeyHashBech32 =
+-- | Encode 'DRepID' as Bech32 with "drep" human readable prefix.
+encodeDRepIDBech32 :: DRepID -> Text
+encodeDRepIDBech32 drepid =
     Bech32.encodeLenient hrp
         . Bech32.dataPartFromBytes
-        . getDRepKeyHash
+        $ appendCip0129BytePrefix
   where
     hrp = [Bech32.humanReadablePart|drep|]
-
--- | Decode a Bech32 encoded 'DRepKeyHash'.
-decodeDRepKeyHashBech32 :: Text -> Either TextDecodingError DRepKeyHash
-decodeDRepKeyHashBech32 t =
+    appendCip0129BytePrefix = case drepid of
+        DRepFromKeyHash (DRepKeyHash payload) ->
+            BS.cons fstByteDRepKeyHash payload
+        DRepFromScriptHash (DRepScriptHash payload) ->
+            BS.cons fstByteDRepScriptHash payload
+
+-- | Decode a Bech32 encoded 'DRepID'.
+decodeDRepIDBech32 :: Text -> Either TextDecodingError DRepID
+decodeDRepIDBech32 t =
     case fmap Bech32.dataPartToBytes <$> Bech32.decodeLenient t of
         Left _ -> Left textDecodingError
         Right (hrp', Just bytes)
-            | hrp' == hrp -> Right $ DRepKeyHash bytes
+            | hrp' == hrp ->
+              let (fstByte, payload) = first BS.head $ BS.splitAt 1 bytes
+              in case fstByte of
+                  0b00100010 ->
+                      Right $ DRepFromKeyHash (DRepKeyHash payload)
+                  0b00100011 ->
+                      Right $ DRepFromScriptHash (DRepScriptHash payload)
+                  _ ->
+                      Left textFirstByteError
         Right _ -> Left textDecodingError
       where
         textDecodingError = TextDecodingError $ unwords
             [ "Invalid DRep key hash: expecting a Bech32 encoded value"
             , "with human readable part of 'drep'."
             ]
-        hrp = [Bech32.humanReadablePart|drep|]
-
--- | Encode 'DRepScriptHash' as Bech32 with "drep_script" hrp.
-encodeDRepScriptHashBech32 :: DRepScriptHash -> Text
-encodeDRepScriptHashBech32 =
-    Bech32.encodeLenient hrp
-        . Bech32.dataPartFromBytes
-        . getDRepScriptHash
-  where
-    hrp = [Bech32.humanReadablePart|drep_script|]
-
--- | Decode a Bech32 encoded 'DRepScriptHash'.
-decodeDRepScriptHashBech32 :: Text -> Either TextDecodingError DRepScriptHash
-decodeDRepScriptHashBech32 t =
-    case fmap Bech32.dataPartToBytes <$> Bech32.decodeLenient t of
-        Left _ -> Left textDecodingError
-        Right (hrp', Just bytes)
-            | hrp' == hrp -> Right $ DRepScriptHash bytes
-        Right _ -> Left textDecodingError
-      where
-        textDecodingError = TextDecodingError $ unwords
-            [ "Invalid DRep Script hash: expecting a Bech32 encoded value"
-            , "with human readable part of 'drep_script'."
+        textFirstByteError = TextDecodingError $ unwords
```

Comment:
Do we want to have concrete error messages in a `Primitive` library? Perhaps it would be better to have some type and let consumers define their own messages?